Adds 'faas-cli template pull' command

ericstoekl opened the pull request:
Pull language templates from any repo that has a 'template/' directory
in the root with 'faas-cli template pull <github repo>'.

Also allows 'faas-cli new --list' to list available languages

Signed-off-by: Eric Stoekl <ems5311@gmail.com>
Signed-off-by: Minh-Quan TRAN <account@itscaro.me>

<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
This PR is to add the ability to use `faas-cli template pull <github repo>`. For example:

```
$ faas-cli template pull https://github.com/itscaro/openfaas-template-php
Fetch templates from repository: https://github.com/itscaro/openfaas-template-php
2017/11/02 21:06:55 HTTP GET https://github.com/itscaro/openfaas-template-php/archive/master.zip
2017/11/02 21:06:56 Writing 8Kb to master.zip

2017/11/02 21:06:56 Attempting to expand templates from master.zip
2017/11/02 21:06:56 Fetched 2 template(s) : [php php5] from https://github.com/itscaro/openfaas-template-php
2017/11/02 21:06:56 Cleaning up zip file...
$ ls template/php5
composer.json  Dockerfile  function  index.php  php-extension.sh  template.yml
$ faas-cli new php_t1 --lang php
Folder: php_t1 created.
  ___                   _____           ____  
 / _ \ _ __   ___ _ __ |  ___|_ _  __ _/ ___| 
| | | | '_ \ / _ \ '_ \| |_ / _` |/ _` \___ \ 
| |_| | |_) |  __/ | | |  _| (_| | (_| |___) |
 \___/| .__/ \___|_| |_|_|  \__,_|\__,_|____/ 
      |_|                                     


Function created in folder: php_t1
Stack file written: php_t1.yml
$ cat php_t1.yml 
provider:
  name: faas
  gateway: http://localhost:8080

functions:
  php_t1:
    lang: php
    handler: ./php_t1
    image: php_t1
$ ls php_t1
composer.json  Handler.php  php-extension.sh
```

Check this asciicinema to see it in action: https://asciinema.org/a/ieyznKidFlvimpeiS5BFMG3ep

Here's a breakdown of what was changed:

### Main Files
- `builder/build.go` - Uses `IsValidTemplate()` to tell whether or not it can build the given language
- `commands/build.go` - Update to pass `overwrite` parameter to `fetchTemplates()`
- `commands/deploy.go` - Get Fprocess from the `template/<lang>/template.yml` file instead of hard-coded
- `commands/fetch_templates.go` - `fetchTemplates()` method pulls from `openfaas/faas-cli` github repo by default, custom repo URL can be passed in. Implements `--overwrite` functionality
- `commands/new_function.go` - Lists available languages from the template dir with `--list` option
- `commands/template_pull.go` - Make sure the passed-in valid, then passes it to `fetchTemplates()`
- `proxy/deploy.go - Remove hard-coded validation of available language templates
- `stack/language_template.go` - Adds `ParseYAMLForLanguageTemplate()`, which will parse the `template/<lang>/template.yml` file, which contains info like Fprocess
- `stack/schema.go` - Adds `LanguageTemplate` type
- `guide/TEMPLATE.md` - Adds a guide for usage of `faas-cli template pull`
- 
  - `template/csharp/template.yml`
  - `template/go-armhf/template.yml`
  - `template/go/template.yml`
  - `template/node-arm64/template.yml`
  - `template/node-armhf/template.yml`
  - `template/node/template.yml`
  - `template/python-armhf/template.yml`
  - `template/python/template.yml`
  - `template/python3/template.yml`
  - `template/ruby/template.yml`

The rest of the files are for testing, and are discussed below.


## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
- [X] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))
First discussed in issue #85, originally work was done in #87.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
In addition to the following files being added for testing, @BurtonR reported success when manually testing: https://github.com/openfaas/faas-cli/pull/87#issuecomment-341608080

 - `commands/fetch_templates_test.go` - Simplify tests, add test for `fetchTemplates()`
 - `commands/new_function_test.go` - Test `--list` option, and add more tests for `new`
 - `commands/template_pull_test.go` - Test `template pull` functionality
 - `commands/testdata/master_test.zip` - Minimal template zip for testing, served by the mock test server
 - `commands/testdata/new_function/template/csharp/function/.gitignore` - Add directories
 - `commands/testdata/new_function/template/go-armhf/.gitignore`
 - `commands/testdata/new_function/template/go/function/.gitignore`
 - `commands/testdata/new_function/template/node-arm64/function/.gitignore`
 - `commands/testdata/new_function/template/node-armhf/.gitignore`
 - `commands/testdata/new_function/template/node/function/.gitignore`
 - `commands/testdata/new_function/template/python-armhf/.gitignore`
 - `commands/testdata/new_function/template/python/function/.gitignore`
 - `commands/testdata/new_function/template/python3/function/.gitignore`
 - `commands/testdata/new_function/template/ruby/function/.gitignore`
 - `stack/language_template_test.go` - Test `ParseYAMLForLanguageTemplate()` against different types of YAML file

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [X] I have signed-off my commits with `git commit -s`
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

Commit:
Adds 'faas-cli template pull' command

Pull language templates from any repo that has a 'template/' directory
in the root with 'faas-cli template pull <github repo>'.

Also allows 'faas-cli new --lang' to list available languages

Signed-off-by: Eric Stoekl <ems5311@gmail.com>
Signed-off-by: Minh-Quan TRAN <account@itscaro.me>

itscaro commented:
@ericstoekl `new --list` instead of `new --lang` in your PR description

alexellis commented:
I've tested the PR and really liked the experience, this will be a great addition to our CLI. Thoughts need to lead on to:

* How do we shrink-wrap templates?
* How do we pin a specific commit/SHA or version of a template?
* What happens at CI time? Do we do `faas-cli template pull`, or store this data alongside stack.yml or inside $HOME/openfaas.yml?

alexellis commented on commands/fetch_templates.go:
```diff
@@ -18,113 +19,161 @@ import (
+		if languageSplit := strings.Split(relativePath, "/"); len(languageSplit) > 2 {
+			language = languageSplit[1]
+
+			if len(languageSplit) == 3 && relativePath[len(relativePath)-1:] == "/" {
```
Pre-squash I asked for these magic numbers to be documented / split out either into a new function or named consts. If I had to work on this I wouldn't know what to do.

alexellis commented on commands/fetch_templates.go:
```diff
@@ -137,3 +186,29 @@ func createPath(relativePath string, perms os.FileMode) error {
 	return err
 }
+
+// Tells whether the language can be expanded from the zip or not
```
Comment should start "canWriteLanguage tells whether"

alexellis commented on commands/fetch_templates.go:
```diff
@@ -137,3 +186,29 @@ func createPath(relativePath string, perms os.FileMode) error {
+func canWriteLanguage(availableLanguages *map[string]bool, language string, overwrite bool) bool {
+	canWrite := false
+	if len(language) > 0 {
+		if _, ok := (*availableLanguages)[language]; ok {
```
Please add `if availableLanguages != nil` and change `ok` to `exists` or `found`

alexellis commented on commands/fetch_templates_test.go:
```diff
@@ -9,16 +9,46 @@ import (
+}
+
+func Test_fetchTemplates(t *testing.T) {
+	defer tearDown_fetch_templates(t)
```
This should be run at the start and at the end - because you may have had an inconsistent file-system if the files were written out and the code panicked and crashed mid-way.

alexellis commented on commands/new_function.go:
```diff
@@ -44,15 +44,23 @@ language or type in --list for a list of languages available.`,
 }
 
 func runNewFunction(cmd *cobra.Command, args []string) {
```
We're changing the commands to return (error) so we can handle the exit codes during CI / scripting. If that's beyond scope of the PR please raise an issue instead to log the work.

ericstoekl replied:
I will take you up on the second option: https://github.com/openfaas/faas-cli/issues/210

alexellis commented on commands/new_function.go:
```diff
@@ -132,14 +141,10 @@ functions:
-		found = true
+func printAvailableTemplates(availableTemplates []string) string {
+	var result string
+	for _, template := range availableTemplates {
```
Can we sort "printAvailableTemplates" deterministically?

alexellis commented on commands/template_pull.go:
```diff
@@ -0,0 +1,80 @@
+// Copyright (c) Alex Ellis, Eric Stoekl 2017. All rights reserved.
```
License should be:

```
// Copyright (c) Alex Ellis 2017. All rights reserved.
```
or
```
// Copyright (c) OpenFaaS Project 2017. All rights reserved.
```

Please see: https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md#license

alexellis commented on commands/template_pull_test.go:
```diff
@@ -0,0 +1,144 @@
+// Copyright (c) Alex Ellis, Eric Stoekl 2017. All rights reserved.
```
https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md#license

alexellis commented on guide/TEMPLATE.md:
```diff
@@ -0,0 +1,56 @@
+# Using template from external repository
```
I wonder if this should be in the guides in the faas repo?

alexellis commented on stack/language_template.go:
```diff
@@ -0,0 +1,61 @@
+	return ParseYAMLDataForLanguageTemplate(fileData)
+}
+
+// iParseYAMLDataForLanguageTemplate parses YAML data into language template
```
Typo of "i" at start of comment.

alexellis commented on stack/schema.go:
```diff
@@ -47,3 +47,8 @@ type Services struct {
 	Provider  Provider            `yaml:"provider,omitempty"`
 }
+
+type LanguageTemplate struct {
```
Comment:

// LanguageTemplate read from template.yml within root of a language template folder

alexellis commented on commands/deploy.go:
```diff
@@ -160,6 +160,28 @@ func runDeploy(cmd *cobra.Command, args []string) {
 			}
 
+			// Get FProcess to use from the ./template/template.yml, if a template is being used
+			if function.Language != "" && function.Language != "Dockerfile" && function.Language != "dockerfile" {
```
Extract to named method.

Change to this or similar:

```
len(function.Language) > 0 && strings.Lower(function.Language) != "dockerfile"
```

alexellis reviewed (changes requested):
Very close to finished. Please address comments and licensing then add a new commit, do not squash.

Commit:
template pull - address issues from thread #201

1) Moves code that merges zip file contents into template/ dir into
expandTemplatesFromZip() function. Removes hard-coded constant.
2) Sort list of languages found with 'faas-cli new --list' before
printing
3) Fixes for typos, more descriptive comments, and remove incorrect
license attribution

Signed-off-by: Eric Stoekl <ems5311@gmail.com>

ericstoekl commented:
Hi @alexellis , thanks for your comments. I've attempted to address all of them except the comment about the guide needing to go into the `faas` repo. Maybe we can have the guide here for now, and move it in a future PR?

I've created a new function in `fetch_templates.go` called `expandTemplatesFromZip()`. The goal is to encapsulate the unwieldy blob of code in (formerly in `fetchTemplates()`) that was there to unpack the template zip file, and merge it with the existing `template/` dir, with the option of overwriting. This function returns 3 items: a list of language templates that could not be overwritten, a list of language templates that were downloaded, and an `error` variable.

I've added sort functionality to the `faas-cli new --list` command, to sort the list of available language templates before displaying them. This is done by implementing the `Len()`, `Swap()`, and `Less()` functions, as is described by the docs: https://golang.org/pkg/sort/

Commit:
template pull - disallow function creation when 'template.yml'
is not present in language template directory

Needs to add a 'template.yml' to each fake language template dir in
commands/testdata/new_function/template/<lang> in order to have current
tests pass. Deletes some unnecessary .gitignore files as a result.

Signed-off-by: Eric Stoekl <ems5311@gmail.com>

ericstoekl commented:
@alexellis , I have pushed another commit to address this comment: https://github.com/openfaas/faas-cli/pull/201#issuecomment-342225124

This is to disallow creation of a function when `template.yml` is not found in the template directory.

I have pushed a new commit to address this issue. If we want to split this off into a future PR, I'm okay with that as well.

alexellis commented on stack/language_template_test.go:
```diff
@@ -0,0 +1,77 @@
+		},
+	}
+
+	for k, i := range dataProvider {
```
I don't know what `dataProvider` means.

alexellis commented on template/node-armhf/template.yml:
```diff
@@ -0,0 +1,2 @@
+language: node-armhf
+fprocess: node index.js
```
Does the node template contain a full template or does it still only have the Dockerfile patch? cc @rgee0

alexellis commented on commands/fetch_templates.go:
```diff
@@ -18,113 +19,179 @@ import (
 			}
+		} else {
+			// template/
+			continue
```
I don't like the half-dozen `continue` statements, how can we re-write this to be clearer?

alexellis commented on commands/fetch_templates.go:
```diff
@@ -18,113 +19,179 @@ import (
+			continue
+		}
+
+		if pathSplit := strings.Split(relativePath, "/"); len(pathSplit) > 2 {
```
Perhaps this all needs to be a function of its own. Extract a new function

alexellis commented on commands/fetch_templates.go:
```diff
@@ -137,3 +204,29 @@ func createPath(relativePath string, perms os.FileMode) error {
+}
+
+// Takes a language input (e.g. "node"), tells whether or not it is OK to download
+func checkLanguage(language string, overwrite bool) bool {
```
`templateFolderExists` / `targetExists`?

alexellis commented on commands/fetch_templates.go:
```diff
@@ -18,113 +19,179 @@ import (
+			isDirectory = relativePath[len(relativePath)-1:] == "/"
+
+			// Check if this is the root directory for a language (at ./template/lang)
+			if len(pathSplit) == rootLanguageDirSplitCount && isDirectory {
```
What are we doing here exactly?

We unzip and then don't overwrite if a `template/node` folder exists?

ericstoekl replied:
Yep, don't overwrite if the `template/<lang>` folder already exists. Unless `--overwrite` flag is active.

Commit:
Updates to 'faas-cli template pull'

1) Improve readability by adding canExpandTemplateData() function, which
returns an enum telling what to do with the data found in the zip
archive.
2) Add function data to ARMHF directories in template/ dir
3) More descriptive naming for test array in language_template_test.go
and checkLanguage() function changed to templateFolderExists()

Signed-off-by: Eric Stoekl <ems5311@gmail.com>

ericstoekl commented:
@alexellis , just made those requested updates. I've refactored the part in the loop with `continue` usage to use a separate function which returns an enum that is then used in a `switch` statement to tell whether to continue with the loop or not. I've also changed the naming of several identifiers as requested, and included the function data into the `go-armhf`, `node-armhf`, and `python-armhf` directories. Let me know what you think. Thanks!

alexellis commented:
Needs a rebase.

alexellis commented:
Rebasing and carrying out some review / clean-up over in #220.
